File: 78-subsets/subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:

        def generate(nums, index, current):
            if len(nums) == index:
                result.append(current[:])
                return result
            
            generate(nums, index + 1, current + [nums[index]])
            generate(nums, index + 1, current)

            return result
            

            


        result = []
        result = generate(nums, 0, [])
        return result

File: 78-subsets/subsets.java
class Solution {

    public void backtrack(List<Integer> current, int[] nums, int index, List<List<Integer>> result){

        int n = nums.length;
        if (index == n) {
            result.add(new ArrayList<>(current));
            return;
        }

        current.add(nums[index]);
        backtrack(current, nums, index + 1, result);
        current.remove(current.size() - 1);
        backtrack(current, nums, index + 1, result);
    }

    public List<List<Integer>> subsets(int[] nums) {

        int n = nums.length;
        List<List<Integer>> result = new ArrayList<>();

        backtrack(new ArrayList<>(), nums, 0, result);

        return result;
    }
}